File: fsharp/QueensHalf.fs
// Copyright 2018 John Hurst
// John Hurst (john.b.hurst@gmail.com)
// 2018-01-28

open System
open System.Diagnostics

type Board = {Size: int; Row: int; Cols: int; Diags1: int; Diags2: int;}

let empty size = {Size = size; Row = 0; Cols = 0; Diags1 = 0; Diags2 = 0}

let ok board col =
    ((board.Cols &&& (1 <<< col)) |||
     (board.Diags1 &&& (1 <<< board.Row + col)) |||
     (board.Diags2 &&& (1 <<< board.Row - col + board.Size - 1))) = 0

let place board col = {
    board with
        Row = board.Row + 1 ;
        Cols = board.Cols ||| (1 <<< col);
        Diags1 = board.Diags1 ||| (1 <<< board.Row + col);
        Diags2 = board.Diags2 ||| (1 <<< board.Row - col + board.Size - 1)
}

let rec solve_rest board =
    if board.Row = board.Size then 1
        else seq { 0 .. board.Size - 1 }
            |> Seq.filter (ok board)
            |> Seq.map (place board)
            |> Seq.map solve_rest
            |> Seq.sum

let solve board =
    // solve for half the columns on first row, double result for symmetry
    let half_cols = seq { 0 .. board.Size / 2 - 1 }
                        |> Seq.map (place board)
                        |> Seq.map solve_rest
                        |> Seq.map (fun v -> 2 * v)
                        |> Seq.sum
    // add middle column if odd number
    let extra_col = if board.Size % 2 = 0 then 0
                        else board.Size / 2
                            |> (place board)
                            |> solve_rest
    half_cols + extra_col

[<EntryPoint>]
let main argv =
    let first = if argv.Length = 0 then 8 else int argv.[0]
    let last = if argv.Length = 1 then first else int argv.[1]
    for size = first to last do
        let watch = Stopwatch.StartNew()
        let count = size |> empty |> solve
        watch.Stop()
        let ts = watch.Elapsed
        let elapsed = (float ts.Hours) * 3600.0 + (float ts.Minutes) * 60.0 + (float ts.Seconds) + (float ts.Milliseconds) / 1000.0
        printfn "%d,%d,%0.3f" size count elapsed
    0

File: cplusplus/QueensHalf.cpp
// Copyright 2018 John Hurst
// John Hurst (john.b.hurst@gmail.com)
// 2018-01-06

#include <iomanip>
#include <iostream>

class Board {
public:
  Board(int size, int row = 0, int cols = 0, int diags1 = 0, int diags2 = 0);
  bool ok(int col);
  Board place(int col);
  int solve_rest();
  int solve();

private:
  int size_;
  int row_ = 0;
  long cols_ = 0;
  long diags1_ = 0;
  long diags2_ = 0;
};

Board::Board(int size, int row, int cols, int diags1, int diags2) {
  size_ = size;
  row_ = row;
  cols_ = cols;
  diags1_ = diags1;
  diags2_ = diags2;
}

bool Board::ok(int col) {
  return ((cols_ & (1 << col)) |
         (diags1_ & (1 << (row_ + col))) |
         (diags2_ & (1 << (row_ - col + size_ - 1)))) == 0;
}

Board Board::place(int col) {
  return Board(
    size_,
    row_ + 1,
    cols_ | 1 << col,
    diags1_ | 1 << (row_ + col),
    diags2_ | 1 << (row_ - col + size_ - 1)
  );
}

int Board::solve_rest() {
  if (row_ == size_) {
    return 1;
  }
  else {
    int result = 0;
    for (int col = 0; col < size_; col++) {
      if (ok(col)) {
        result += place(col).solve_rest();
      }
    }
    return result;
  }
}

int Board::solve() {
  int result = 0;
  // solve for half the columns on first row, double result for symmetry
  for (int col = 0; col < size_ / 2; col++) {
    result += 2 * place(col).solve_rest();
  }
  // add middle column if odd number
  if (size_ % 2 == 1) {
    result += place(size_ / 2).solve_rest();
  }
  return result;
}

int main(int argc, char** argv) {
  int from = argc < 2 ? 8 : atoi(argv[1]);
  int to = argc < 3 ? from : atoi(argv[2]);

  for (int size = from; size <= to; size++) {
    clock_t start = clock();
    int result = Board(size).solve();
    clock_t end = clock();
    float elapsed = (end - start) / 1000000.0;
    std::cout << size << "," << result << "," << std::setprecision(3) << elapsed << std::endl;
  }
}

